lib/rpcs_dpdk/ethdev.c

```diff
@@ -2739,3 +2739,169 @@ TARPC_FUNC(rte_eth_rx_metadata_negotiate,
 done:
     ;
 })
+
+static void
+tarpc_rte_eth_fec_capa_rte2rpc(uint32_t rte, uint32_t *rpc)
+{
+    *rpc = 0;
+
+#define RTE_ETH_FEC_MODE_RTE2RPC(_mode) \
+    do {                                                               \
+        uint32_t capa = RTE_ETH_FEC_MODE_TO_CAPA(RTE_ETH_FEC_##_mode); \
+                                                                       \
+        if (rte & capa)                                                \
+        {                                                              \
+            rte &= ~capa;                                              \
+            *rpc |= (1ULL << TARPC_RTE_ETH_FEC_##_mode);               \
+        }                                                              \
+    } while (0)
+
+    RTE_ETH_FEC_MODE_RTE2RPC(NOFEC);
+    RTE_ETH_FEC_MODE_RTE2RPC(AUTO);
+    RTE_ETH_FEC_MODE_RTE2RPC(RS);
+    RTE_ETH_FEC_MODE_RTE2RPC(BASER);
+
+#undef RTE_ETH_FEC_MODE_RTE2RPC
```

Comment:
I strongly dislike the abuse of macros here.

Reply:
By using a loop over a table, I presume (note the code below is just an idea, not a carved in stone commandment)
```
typedef struct mapping {
     uint32_t native;
     unsigned int te_bit;
} mapping;

static uint32_t native2te(uint32_t native, const mapping *map, unsigned int unknown_bit)
{
      unsigned int i;
      uint32_t result = 0;
      
      for (i = 0; map[i].native != MAPPING_END; i++)
      {
          if ((native & map[i].native) != 0)
          {
               native &= ~map[i].native;
               result |= 1u << map[i].te_bit;
          }
      }
      if (native != 0)
          result |= 1u << unknown_bit;
      return result;    
}
```

Then it's totally ok to use macros to define conversion tables, to eliminate common prefixes:
```
static mapping fec_mode_map[] = {
#define ITEM(mode_) {.native = RTE_ETH_FEC_MODE_TO_CAPA(RTE_ETH_FEC_##mode_, .te_bit = TARPC_RTE_ETH_FEC_##mode_}
ITEM(NOFEC),
ITEM(AUTO),
ITEM(RS),
ITEM(BASER),
#undef ITEM
}
```

Reply:
We have `te_enum.h` which have `te_enum_translate` to convert between enum values. It can be naturally augmented with a function to convert bitmasks. There was even a patch to do that, but it got stale. So you can add it there.